src/Titon/Context/Depository.hh

```diff
@@ -63,6 +71,32 @@ class Depository {
     }
 
     /**
+     * Register a new Service Provider object in the container.
+     *
+     * @param mixed $serviceProvider
+     *
+     * @return $this Return the depository for fluent method chaining
+     */
+    public function addServiceProvider(mixed $serviceProvider): this {
+        if (!($serviceProvider instanceof ServiceProvider)) {
+            $serviceProvider = $this->make($serviceProvider);
+        }
+
+        // Type checker thinks we are `mixed` here instead of `ServiceProvider. Reset it.
+        invariant($serviceProvider instanceof ServiceProvider, 'Must be a ServiceProvider.');
+
+        $serviceProvider->setDepository($this);
+
+        if ($serviceProvider->getProvides()->isEmpty()) {
```

Comment:
When creating a new ServiceProvider, if the user doesn't fill out the `provides` class variable, then we need to go ahead and `register` the ServiceProvider, otherwise we won't know what the object provides. If `provides` is filled out by the user, then we can actually defer the `register` of the provider until we need it, in which case, during `make`, we can call `register` then.